File: trustgraph-base/trustgraph/base/consumer_spec.py
from __future__ import annotations

from typing import Any

from . spec import Spec

class ConsumerSpec(Spec):
    def __init__(self, name: str, schema: Any, handler: Any, concurrency: int = 1) -> None:
        self.name = name
        self.schema = schema
        self.handler = handler
        self.concurrency = concurrency

    async def register(self, flow: Any, processor: Any, definition: dict[str, Any]) -> Any:

        topic = definition["topics"][self.name]
        subscription = (
            processor.id + "--" + flow.workspace + "--" +
            flow.name + "--" + self.name
        )

        handler = self.handler

        async def pool_handler(message):
            await handler(message, None, flow)

        reg = await processor.receiver_pool.add_consumer(
            topic=topic,
            subscription=subscription,
            schema=self.schema,
            handler=pool_handler,
        )

        flow.consumer[self.name] = reg
        return reg

File: ts/packages/base/src/spec/consumer-spec.ts
/**
 * Consumer specification — declares a message consumer for a flow.
 *
 * Python reference: trustgraph-base/trustgraph/base/consumer_spec.py
 */

import { Effect } from "effect";
import type { Spec } from "./types.js";
import type { SpecRuntimeRequirements } from "./types.js";
import type { Flow, FlowDefinition } from "../processor/flow.js";
import {
  ConsumerFactory,
  type EffectMessageHandler,
} from "../messaging/runtime.js";
import {
  type PubSubError,
} from "../errors.js";

declare const ConsumerSpecType: unique symbol;

export interface ConsumerSpec<T, E = never, R = never> extends Spec<R> {
  readonly [ConsumerSpecType]?: {
    readonly message: T;
    readonly error: E;
  };
  readonly addEffect: (
    flow: Flow<R>,
    definition: FlowDefinition,
  ) => Effect.Effect<void, PubSubError, SpecRuntimeRequirements | R>;
}

export function makeConsumerSpec<T, E = never, R = never>(
  name: string,
  handler: EffectMessageHandler<T, E, R>,
  concurrency = 1,
): ConsumerSpec<T, E, R> {
  const addEffect = Effect.fn("ConsumerSpec.addEffect")(function* (
    flow: Flow<R>,
    definition: FlowDefinition,
  ) {
      const topic = definition.topics?.[name] ?? name;
      const factory = yield* ConsumerFactory;
      const consumer = yield* factory.run<T, E, R>(
        {
          topic,
          subscription: `${flow.processorId}-${flow.name}-${name}`,
          handler,
          concurrency,
        },
        { id: flow.processorId, name: flow.name, flow },
      );
      flow.registerConsumer(name, consumer);
  });

  return {
    name,
    addEffect,
  };
}
